src/Translation.cpp

```diff
@@ -0,0 +1,107 @@
+#include "Translation.h"
+#include "3rd_party/bergamot-translator/src/translator/response.h"
+#include <QDebug>
+
+namespace {
+
+bool contains(marian::bergamot::ByteRange const &span, std::size_t offset) {
+    return offset >= span.begin && offset < span.end;
+}
+
+auto findWordByByteOffset(marian::bergamot::Annotation const &annotation, std::size_t pos) {
+    std::size_t sentenceIdx, wordIdx;
+
+    for (sentenceIdx = 0; sentenceIdx < annotation.numSentences() - 1; ++sentenceIdx)
+        if (annotation.sentence(sentenceIdx).end > pos) // annotations are sorted, so this should be safe to assume.
+            break;
+
+    // if not found, sentenceIdx == numSentences() - 1 (i.e. the last sentence)
+    
+    for (wordIdx = 0; wordIdx < annotation.numWords(sentenceIdx) - 1; ++wordIdx)
+        if (annotation.word(sentenceIdx, wordIdx).end >= pos)
+            break;
+
+    // Same. worst case, wordIdx == annotation.numWords(sentenceIdx) - 1
+
+    return std::tuple(sentenceIdx, wordIdx);
+}
+
+int offsetToPosition(std::string const &text, std::size_t offset) {
+    if (offset > text.size())
+        return -1;
+
+    return QString::fromUtf8(text.data(), offset).size();
+}
+
+std::size_t positionToOffset(std::string const &text, int pos) {
+    return QString::fromStdString(text).left(pos).toLocal8Bit().size();
```

Comment:
Maybe change here to use `fromUtf8`?

Reply:
In this case I need `QString::left()` because of char pos. The `size` argument of `fromUtf8` won't do since that's bytes.

I checked whether there would be any other benefit for using `fromUtf8`, but `fromStdString` appears to be identical to `fromUtf8(text.data(), int(text.size()))`.

Reply:
Hmm... It seems all of them do a copy of the memory ;/

Reply:
`fromRawData` seems to be the only zero-copy solution, but as we discussed it requires `QChar`...

Reply:
I've replaced the code with something based on https://stackoverflow.com/a/3586973.